bundles/org.openhab.binding.myuplink/src/main/java/org/openhab/binding/myuplink/internal/discovery/MyUplinkDiscoveryService.java

```diff
@@ -0,0 +1,144 @@
+import org.openhab.binding.myuplink.internal.handler.MyUplinkAccountHandler;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * this class will handle discovery of wallboxes and circuits within the site configured.
+ *
+ * @author Alexander Friese - initial contribution
+ *
+ */
+@NonNullByDefault
+public class MyUplinkDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(MyUplinkDiscoveryService.class);
+    private @NonNullByDefault({}) MyUplinkAccountHandler bridgeHandler;
+
+    public MyUplinkDiscoveryService() throws IllegalArgumentException {
+        super(Set.of(MyUplinkBindingConstants.THING_TYPE_ACCOUNT), 300, false);
+    }
+
+    @Override
+    public void setThingHandler(ThingHandler handler) {
+        if (handler instanceof MyUplinkAccountHandler accountHandler) {
+            this.bridgeHandler = accountHandler;
+            this.bridgeHandler.setDiscoveryService(this);
+        }
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return bridgeHandler;
+    }
+
+    // method is defined in both implemented interface and inherited class, thus we must define a behaviour here.
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    @Override
+    protected void startScan() {
+        bridgeHandler.enqueueCommand(new GetSystems(bridgeHandler, this::processMyUplinkDiscoveryResult));
+    }
+
+    /**
+     * callback that handles json result data to provide discovery result.
+     *
+     * @param site
+     */
+    void processMyUplinkDiscoveryResult(CommunicationStatus status, JsonObject json) {
+        logger.debug("processMyUplinkDiscoveryResult {}", json);
+
+        JsonArray systems = json.getAsJsonArray(JSON_KEY_SYSTEMS);
+        if (systems == null || systems.isEmpty()) {
+            logger.info("System discovery failed, no systems found.");
+        } else {
+            systems.forEach(this::handleSystemDiscovery);
+        }
+    }
+
+    void handleSystemDiscovery(JsonElement json) {
+        logger.debug("handleSystemDiscovery {}", json);
+
+        JsonObject system = json.getAsJsonObject();
+        String systemId = Utils.getAsString(system, JSON_KEY_SYSTEM_ID);
+        JsonArray devices = system.getAsJsonArray(JSON_KEY_DEVICES);
+        if (devices == null || devices.isEmpty()) {
+            logger.info("System discovery failed, no devices found.");
```

Comment:
The discovery proces is considered succesfull if it finishes wihout issues. Regardless if a device is found or not.

Reply:
fixed